src/NetworkResolver/NetworkResolverInterface.php

```diff
@@ -0,0 +1,18 @@
+<?php
+
+
+namespace Yiisoft\Yii\Web\NetworkResolver;
+
+use Psr\Http\Message\ServerRequestInterface;
+
+interface NetworkResolverInterface
+{
+    public function getRemoteIp(): string;
```

Comment:
It cannot work without server request so either `ServerRequestInterface` should be injected from constructor or passed as a method argument. I prefer:

```php
 public function getRemoteIp(ServerRequestInterface $serverRequest): string;
```

Reply:
> else use $serverRequest from the object ($this)

Where would you get it from? It's not in the container.

Reply:
> Where would you get it from? It's not in the container.

```php
class AnyNetworkResolver implements... {
  private $serverRequest;
  public function withServerRequest($serverRequest) {...}
```

The object can be used multiple times during runtime. Under normal use, it must be instantiated the first time the `$serverRequest` object appears. However, the instance should be accessible anywhere where the `$serverRequest` object is accessible.

User IP usage examples:
- insert into database
- condition in controller
- display in view

Based on the tests, I can see the current architecture of the system and not see where it can be injected.

```php
return new MiddlewareDispatcher([
    $errorCatcher,
    $networkResolver, // <---- It should not be placed in a parameter here, but rather in a `$container`.
    $session,
    $router, // controller would get Request that was modified by networkResolver.
                 // No need to pass networkResolver as dependency of the widget
                 //   ^----------------- It's OK
], $container);
//     ^
//     \-------------------- Maybe we should store it here?
```

Reply:
Yes, NetworkResolver could be put into container but it won't get request instance anyway. So why having `withRequest()` then? I think this method should be removed.

Reply:
> So why having withRequest() then? I think this method should be removed.

The `$request` object passed to all methods is a bit like a static helper class. I think you need to find the right point in the code where
- `$request` object already exists
- At the configuration level has already been added to the container (network resolver)

Unless a middle ware setting something in the network resolver, its output does not change during the web request. If the object internally stores the request object, it can cache the output results. However, if you get the request object in a parameter, I cannot assume that the request object will not change.

Reply:
Final result of planning:
- `Middleware\NetworkResolver`: If needed, it modifies the request
- `NetworkResolver\NetworkResolverInterface`: resolving the network. Depending on the use, may be placed in a container or helper (if needed, eg user ip).

Reply:
OK.